src/io.c

```diff
@@ -0,0 +1,2024 @@
+    Pr("%c", (Int)'\03', 0L);
+    if (!STATE(Output)->isstream) {
+        SyFclose(STATE(Output)->file);
+    }
+
+/* revert to previous output file and indicate success                 */
+#ifdef HPCGAP
+    const int sp = --STATE(OutputFilesSP);
+    STATE(Output) = sp ? STATE(OutputFiles)[sp - 1] : 0;
+#else
+    STATE(Output)--;
+#endif
+    return 1;
+}
+
+
+/****************************************************************************
+**
+*F  OpenAppend( <filename> )  . . open a file as current output for appending
+**
+**  'OpenAppend' opens the file  with the name  <filename> as current output.
+**  All subsequent output will go  to that file, until either   it is  closed
+**  again  with 'CloseOutput' or  another  file is  opened with 'OpenOutput'.
+**  Unlike 'OpenOutput' 'OpenAppend' does not truncate the file to size 0  if
+**  it exists.  Appart from that 'OpenAppend' is equal to 'OpenOutput' so its
+**  description applies to 'OpenAppend' too.
+*/
+UInt OpenAppend(const Char * filename)
+{
+    Int file;
+
+    /* fail if we can not handle another open output file                  */
+    if (STACK_SIZE(Output) == ARRAY_SIZE(STATE(OutputFiles)))
+        return 0;
+
+#ifdef HPCGAP
+    if (!strcmp(filename, "*defout*"))
+        return OpenDefaultOutput();
+#endif
+
+    /* try to open the file                                                */
+    file = SyFopen(filename, "a");
+    if (file == -1)
+        return 0;
+
+/* put the file on the stack, start at position 0 on an empty line     */
+#ifdef HPCGAP
+    const int sp = STATE(OutputFilesSP)++;
+    if (!STATE(OutputFiles)[sp]) {
+        STATE(OutputFiles)[sp] = NewOutput();
+    }
+    STATE(Output) = STATE(OutputFiles)[sp];
+#else
+    assert(STATE(Output) != 0);
+    STATE(Output)++;
+#endif
+    STATE(Output)->file = file;
+    STATE(Output)->line[0] = '\0';
+    STATE(Output)->pos = 0;
+    STATE(Output)->indent = 0;
+    STATE(Output)->isstream = 0;
+
+    /* variables related to line splitting, very bad place to split        */
+    STATE(Output)->hints[0] = -1;
+
+    /* indicate success                                                    */
+    return 1;
+}
+
+
+/****************************************************************************
+**
+*F * * * * * * * * * * * * * * input functions  * * * * * * * * * * * * * * *
+*/
+
+
+/****************************************************************************
+**
+*V  ReadLineFunc  . . . . . . . . . . . . . . . . . . . . . . . .  'ReadLine'
+*/
+
+
+/****************************************************************************
+**
+*F  GetLine2( <input>, <buffer>, <length> ) . . . . . . . . get a line, local
+*/
+static Int GetLine2(TypInputFile * input, Char * buffer, UInt length)
+{
+#ifdef HPCGAP
+    if (!input) {
+        input = STATE(Input);
+        if (!input)
+            OpenDefaultInput();
+        input = STATE(Input);
+    }
+#endif
+
+    if (input->isstream) {
```

Comment:
Off-topic (resp. future topic): this is something else we could refactor: instead of having special cases for `isstream`, make this generic, by having function pointers inside `TypInputFile` and `TypOutputFile` for things like `read`, `write`, etc.. Then besides the current "file types", we can easily create more, e.g. a proper GZip input/output stream (w/o relying on an external gzip process), and more.